File: app/src/main/java/me/zhanghai/android/files/settings/BookmarkDirectoriesPreference.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.settings;

import android.content.Context;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.widget.TextView;

import java.util.List;

import androidx.annotation.AttrRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StyleRes;
import androidx.lifecycle.Observer;
import androidx.preference.Preference;
import androidx.preference.PreferenceViewHolder;
import me.zhanghai.android.files.compat.ListFormatterCompat;
import me.zhanghai.android.files.navigation.BookmarkDirectory;
import me.zhanghai.java.functional.Functional;

public class BookmarkDirectoriesPreference extends Preference {

    private final Observer<List<BookmarkDirectory>> mObserver = this::onBookmarkDirectoriesChanged;

    @Nullable
    private CharSequence mSummaryEmpty;

    public BookmarkDirectoriesPreference(@NonNull Context context) {
        super(context);

        init();
    }

    public BookmarkDirectoriesPreference(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);

        init();
    }

    public BookmarkDirectoriesPreference(@NonNull Context context, @Nullable AttributeSet attrs,
                                         @AttrRes int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        init();
    }

    public BookmarkDirectoriesPreference(@NonNull Context context, @Nullable AttributeSet attrs,
                                         @AttrRes int defStyleAttr, @StyleRes int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);

        init();
    }

    private void init() {
        setPersistent(false);
        mSummaryEmpty = getSummary();
    }

    @Override
    public void onAttached() {
        super.onAttached();

        Settings.BOOKMARK_DIRECTORIES.observeForever(mObserver);
    }

    @Override
    public void onDetached() {
        super.onDetached();

        Settings.BOOKMARK_DIRECTORIES.removeObserver(mObserver);
    }

    private void onBookmarkDirectoriesChanged(
            @NonNull List<BookmarkDirectory> bookmarkDirectories) {
        List<String> names = Functional.map(bookmarkDirectories, BookmarkDirectory::getName);
        CharSequence summary = !names.isEmpty() ? ListFormatterCompat.format(names) : mSummaryEmpty;
        setSummary(summary);
    }

    @Override
    public void onBindViewHolder(PreferenceViewHolder holder) {
        super.onBindViewHolder(holder);

        TextView summaryText = (TextView) holder.findViewById(android.R.id.summary);
        summaryText.setSingleLine(true);
        summaryText.setEllipsize(TextUtils.TruncateAt.END);
    }

    @Override
    protected void onClick() {
        Context context = getContext();
        context.startActivity(BookmarkDirectoriesActivity.newIntent(context));
    }
}

File: app/src/main/java/me/zhanghai/android/files/settings/BookmarkDirectoriesPreference.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.settings

import android.content.Context
import android.text.TextUtils
import android.util.AttributeSet
import android.widget.TextView
import androidx.annotation.AttrRes
import androidx.annotation.StyleRes
import androidx.lifecycle.Observer
import androidx.preference.Preference
import androidx.preference.PreferenceViewHolder
import me.zhanghai.android.files.compat.ListFormatterCompat
import me.zhanghai.android.files.navigation.BookmarkDirectory
import me.zhanghai.android.files.util.createIntent
import me.zhanghai.android.files.util.startActivitySafe

class BookmarkDirectoriesPreference : Preference {
    private var emptySummary = summary

    private val observer = Observer<List<BookmarkDirectory>> { onBookmarkDirectoryListChanged(it) }

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet?, @AttrRes defStyleAttr: Int) : super(
        context, attrs, defStyleAttr
    )

    constructor(
        context: Context,
        attrs: AttributeSet?,
        @AttrRes defStyleAttr: Int,
        @StyleRes defStyleRes: Int
    ) : super(context, attrs, defStyleAttr, defStyleRes)

    init {
        isPersistent = false
    }

    override fun onAttached() {
        super.onAttached()

        Settings.BOOKMARK_DIRECTORIES.observeForever(observer)
    }

    override fun onDetached() {
        super.onDetached()

        Settings.BOOKMARK_DIRECTORIES.removeObserver(observer)
    }

    private fun onBookmarkDirectoryListChanged(bookmarkDirectories: List<BookmarkDirectory>) {
        val names = bookmarkDirectories.map { it.name }
        val summary = if (names.isNotEmpty()) ListFormatterCompat.format(names) else emptySummary
        setSummary(summary)
    }

    override fun onBindViewHolder(holder: PreferenceViewHolder) {
        super.onBindViewHolder(holder)

        val summaryText = holder.findViewById(android.R.id.summary) as TextView
        summaryText.ellipsize = TextUtils.TruncateAt.END
        summaryText.isSingleLine = true
    }

    override fun onClick() {
        context.startActivitySafe(BookmarkDirectoryListActivity::class.createIntent())
    }
}
